### PR TITLE
ReadOnly UserDao support

### PR DESCRIPTION
**Pull request by frett**
add the ability to make a UserDao readonly.

### CODE REVIEW

**Comment by frett** [comment]
@hlbraddock 

I created the DaoException class to operate the same as UserException. I kept them separate because in my mind UserExceptions are missing/invalid users and DaoExceptions are errors at the DAO layer. Unfortunately this changes the signature of several methods on the interfaces, which means updates in client code.

**Comment by frett** [comment]
I also left readOnly accessors off of the UserDao interface because I didn't see any current/immediate need for testing the state.

This is going to be part of my disable Self Service functionality for migration and post migration for maintenance if needed.

File: ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoTest.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package org.ccci.idm.user.ldaptive.dao;
+
+import org.ccci.idm.user.dao.AbstractUserDaoTest;
+
+public class LdaptiveUserDaoTest extends AbstractUserDaoTest {
```
Comment (hlbraddock) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoTest.java]: Is this class being used?

Comment (frett) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoTest.java]: it's a test class that inherits the tests from the AbstractUserDaoTest class. That's where I have the readOnly test defined at

Comment (hlbraddock) [inline ldaptive/src/test/java/org/ccci/idm/user/ldaptive/dao/LdaptiveUserDaoTest.java]: Never mind. I see now that it is within `AbstractUserDaoTest.java`.

**Comment by hlbraddock** [comment]
It looks like you are asserting writable (`assertWritable()`) everywhere necessary within `LdaptiveUserDao` and that you are testing all the cases with `AbstractUserDaoTest`.

Looks good.

**Comment by frett** [comment]
yeah, I've been keeping some of the tests/implementation details separate in case we do need to add support down the line for a different implementation. It will let us quickly test that it meets our pre-existing tests.

**Comment by frett** [comment]
my biggest concern is the change of the method signatures, it means retouching code wherever those methods are used.

**Comment by hlbraddock** [comment]
True.

**Comment by frett** [comment]
merging this

**Comment by frett** [comment]
@dsgoers you should be aware of the changing method signatures in this PR

**Pull request merged**